### DISCUSSION TITLE
Can someone explain the difference between Submodules and Remotes?

### DISCUSSION THREAD

**Discussion by DataJuggler** [community/community#22550] [New to GitHub]
<p>A couple of days ago an admin here answered my post about sharing repos and suggest sub modules might be the way to do it.</p>
<p>I watched a few videos on Submodules, and they also had a video on Remotes, and they seem exactly the same to me from the descriptions. Can someone explain the difference between the two, and when would you want to use a remote versus when to use a sub module?</p>
<p>I am leaning towards Azure DevOps over GitHub. Microsoft has too many options that do the same thing.</p>
<p>Thanks</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>They’re actually very different.</p>
<p><a href="https://git-scm.com/book/en/v2/Git-Basics-Working-with-Remotes">Remotes</a> are names that you give to remote locations where the repository is stored. So when you clone a repository from GitHub to your local machine, the repository on GitHub is given a remote named <code>origin</code> in your local repository. This allows you to <code>fetch</code>, <code>push</code>, or <code>pull</code> your changes between your local repository and the copy that is stored on GitHub.</p>
<p><a href="https://git-scm.com/book/en/v2/Git-Tools-Submodules">Submodules</a> are for when you need to compose some other repository <em>inside</em> the main repository. For example, you’re working on a project that does some amazing thing. Then you create some library for this project that you want to store in a separate repository. You would create a separate repository for the library but then include that library into the main project as a submodule so that the entire project can be built all at once.</p>
<p>Virtually every Git repository has at least one remote. Very few projects use submodules if they are using modern dependency management tools</p>
<p>Hopefully this picture will help too …</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/7/753e3fca95606224991652045d06625e816017fa.png" data-download-href="/uploads/short-url/gJbeoZ3TGWVotCKLiyCIW7cNuUq.png?dl=1" title="remotes-vs-submodules.png"><img src="https://user-images.githubusercontent.com/1038121/181071952-4c9e8f7a-f34f-4f7b-8639-820dd5b8b0fb.png" alt="remotes-vs-submodules.png" data-base62-sha1="gJbeoZ3TGWVotCKLiyCIW7cNuUq" width="659" height="500" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/7/753e3fca95606224991652045d06625e816017fa_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">remotes-vs-submodules.png</span><span class="informations">688×522 6.52 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by DataJuggler** [comment]
<p>This is not the type of remotes I am talking about.</p>
<p>I watched a video where a repository had shared Remotes that it used as references.</p>
<p>Either way, I don’t care. Git is too complicated to share all my projects. I just want to upload code.</p>
